### PR TITLE
Add rake task to build a YAML file with list of Azure regions

### PR DESCRIPTION
**Pull request by agrare**
Use the `az` cli to get a current list of Azure regions via the API.

First login with `az login` then run `rake azure:regions:update` to update the list of regions or `rake azure:regions:list` to list them

### CODE REVIEW

**Comment by agrare** [comment]
```ruby
>> old_regions = ManageIQ::Providers::Azure::Regions.regions
=> {"australiaeast"=>{:name=>"australiaeast", :description=>"Australia East"}, "australiacentral"=>{:name=>"australiacentral", :description=>"Australia Central"}, "australiacentral2"=>{:name=>"australiacent...
>> new_regions = YAML.load_file("db/fixtures/azure_regions.yml")
=> {"eastus"=>{:name=>"eastus", :description=>"East US"}, "eastus2"=>{:name=>"eastus2", :description=>"East US 2"}, "southcentralus"=>{:name=>"southcentralus", :description=>"South Central US"}, "westus2"=>...
>> old_regions.keys - new_regions.keys
=> ["germanycentral", "germanynortheast", "usgovarizona", "usgoviowa", "usgovtexas", "usgovvirginia"]
>> new_regions.keys - old_regions.keys
=> ["westus3", "swedencentral", "southafricanorth", "jioindiawest", "norwayeast", "switzerlandnorth", "uaenorth", "centralusstage", "eastusstage", "eastus2stage", "northcentralusstage", "southcentralusstage", "westusstage", "westus2stage", "asia", "asiapacific", "australia", "brazil", "canada", "europe", "france", "germany", "global", "india", "japan", "korea", "norway", "southafrica", "switzerland", "uae", "uk", "unitedstates", "eastasiastage", "southeastasiastage", "centraluseuap", "eastus2euap", "southafricawest", "jioindiacentral", "norwaywest", "switzerlandwest", "uaecentral", "brazilsoutheast"]
```

File: db/fixtures/azure_regions.yml [hunk 1]
```diff
@@ -0,0 +1,223 @@
+---
+eastus:
```
Comment (Fryguy) [inline db/fixtures/azure_regions.yml:2]: Can we sort this list be key?  That way diffs are clearer in the future.

File: lib/tasks_private/azure.rake [hunk 2]
```diff
@@ -1,24 +1,28 @@
...
+
+      regions = JSON.parse(stdout).map do |region|
+        {
+          :name        => region["name"],
+          :description => region["displayName"]
+        }
       end
-      sh "bundle exec rspec #{spec_file}"
+
+      regions_by_name = regions.sort_by { |r| r[:name]}.index_by { |r| r[:name] }
```
Comment (Fryguy) [inline lib/tasks_private/azure.rake:24]:
```suggestion
      regions_by_name = regions.sort_by { |r| r[:name] }.index_by { |r| r[:name] }
```

**Review by Fryguy — APPROVED** [review]

**Comment by djberg96** [comment]
If memory serves, available regions vary depending on your subscription. I vaguely recall jumping through hoops to test Germany back in https://github.com/ManageIQ/manageiq-providers-azure/pull/98.

**Comment by agrare** [comment]
Probably going to have to patch in some regions manually if they aren't returned (e.g. the govcloud ones)

File: db/fixtures/azure_regions.yml [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+  :description: France South
+germany:
+  :name: germany
+  :description: Germany
+germanynorth:
+  :name: germanynorth
+  :description: Germany North
+germanywestcentral:
+  :name: germanywestcentral
+  :description: Germany West Central
```
Comment (Fryguy) [inline db/fixtures/azure_regions.yml:94]: old one was:

```
        "germanycentral"     => {
          :name        => "germanycentral",
          :description => _("Germany Central"),
        },
        "germanynorth"       => {
          :name        => "germanynorth",
          :description => _("Germany North"),
        },
        "germanynortheast"   => {
          :name        => "germanynortheast",
          :description => _("Germany Northeast"),
        },
        "germanywestcentral" => {
          :name        => "germanywestcentral",
          :description => _("Germany West Central"),
        },
```

File: spec/models/manageiq/providers/azure/regions_spec.rb [hunk 4]
```diff
@@ -19,7 +19,7 @@ def azure_regions
     end
 
     it "returns the expected array for the names method" do
-      expect(azure_regions).to match_array(described_class.names)
+      expect(described_class.names).to include(*azure_regions)
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/regions_spec.rb:22]: Not sure we need this to match an exactly list if we're auto-generating it

**Comment by agrare** [comment]
Okay govcloud and those two germany regions are now being merged in as part of running the rake task.

File: lib/tasks_private/azure.rake [hunk 5]
```diff
@@ -1,24 +1,37 @@
...
+
+      regions_by_name = regions.sort_by { |r| r[:name] }.index_by { |r| r[:name] }
+      File.write("db/fixtures/azure_regions.yml", regions_by_name.to_yaml)
+    end
+
+    private
+
+    # These are regions that are not returned by list-locations using a subscription that we have access to
+    def additional_regions
+      [{:name => "germanycentral", :description => "Germany Central"}, {:name => "germanynortheast", :description => "Germany Northeast"}, {:name => "usgovarizona", :description => "US Gov Arizona"}, {:name => "usgoviowa", :description => "US Gov Iowa"}, {:name => "usgovtexas", :description => "US Gov Texas"}, {:name => "usgovvirginia", :description => "US Gov Virginia"}]
```
Comment (Fryguy) [inline lib/tasks_private/azure.rake:34]: Only thing I might suggest is to blast this out 1-per-line, sorted, to simplify future diffs, or perhaps move out to a yml file.

**Comment by Fryguy** [comment]
@agrare A conflict occurred during the backport of this pull request to `morphy`.

If this pull request is based on another pull request that has not been marked for backport, add the appropriate labels to the other pull request. Otherwise, please create a new pull request direct to the `morphy` branch in order to resolve this.

Conflict details:

```diff
* Unmerged path lib/tasks_private/azure.rake
```

**Comment by agrare** [comment]
@Fryguy this backports cleanly with https://github.com/ManageIQ/manageiq-providers-azure/pull/463 applied

**Comment by Fryguy** [comment]
Backported to `morphy` in commit 955537c6b7c80e8cd021dcf7f5b43154a6d3effe.

```text
commit 955537c6b7c80e8cd021dcf7f5b43154a6d3effe
Author: Jason Frey <fryguy9@gmail.com>
Date:   Mon Nov 22 13:25:00 2021 -0500

    Merge pull request #482 from agrare/add_uae_regions
    
    Add rake task to build a YAML file with list of Azure regions
    
    (cherry picked from commit 8daa5e980781c4f902212d808b86d36d0ad3bf9f)
```

**Pull request merged**